doc/sphinx/proof-engine/ssreflect-proof-language.rst

```diff
@@ -4008,8 +4110,19 @@ definition.
       rewrite /=.
       unlock lid.
 
-.. tacn:: unlock {? @occ_switch } @ident
-   :name: unlock
+.. tacn:: unlock {* {? %{ @clear_switch %} } @term } {? @ssr_in }
+   :name: unlock (ssreflect)
+
+   .. insertprodn ssrclausehyps lcpattern
+
+   .. prodn::
+      ssrclausehyps ::= @gen_item {* {? , } @gen_item }
+      gen_item ::= @ssrclear
+      | {? @ } @ident
+      | ( @ident {? := @lcpattern } )
+      | (@ @ident := @lcpattern )
+      ssrclear ::= %{ {+ @ident } %}
+      lcpattern ::= @term
```

Comment:
I think lcpattern is parsed as a term, but I'd like it to be documented with a bit more structure (which is encoded as a term, but this should be hidden).

```
lcpattern := @term | ( @ident in @term ) | ( @term as @ident in @term )
```
Examples: `_ + _` , `(X in X = _)`, `(_ * _) as X in _ + X = _`

Reply:
the patterns for rewrite, between `[...]` are the same (up to enclosing `()`) plus 2 extras
```
rewrite_pattern := [ @term ] | [ in @term ] | [ @ident in @term ] | [ in @ident in @term ] | [ @term as @ident in @term ]
```
If possible the grammar should suggest that lcpattern is a subset of rewrite_pattern, so maybe one should hoist the different parentheses

Reply:
I think the "in" and "as" you refer to are part of `r_pattern` (`rpattern` in the mlgs):

![image](https://user-images.githubusercontent.com/1253341/98177535-b0d82f00-1eaf-11eb-8d1c-ca57b8ba7c8a.png)

Reply:
It's not so easy to follow your comments--you're not mentioning details like "rename r_pattern to be rewrite_pattern" and it takes me 10 minutes to guess that you might have meant that.

> If possible the grammar should suggest that lcpattern is a subset of rewrite_pattern, so maybe one should hoist the different parentheses

`lcpattern` is not a subset of `rewrite_pattern`.  What do you mean by "hoist the different parentheses"?

Reply:
In ssr there is one language of patterns. When a pattern is used to drive rewrite as in `rewrite [pattern]rule` then the pattern is always between `[]`. The same patterns are also available in, for example, `set name := (pattern)` or `move: (pattern)`. There the patterns are between `()`.

The patterns that begin with `in` only make sense for rewrite.

For "technical reasons" sometimes these patterns are parsed as regular terms. One reason is that `move: (term)` is also a thing. Another one, maybe more interesting, is that some "funky" terms can be defined by a notation and used as patterns. For example `RHS` is an abbreviation for `X in _ = X` (the right hand side of an equation). `RHS` is accepted both between `[]` (for rewrite) and between `()` for `set` and `move:`. I'd like the syntax of `move:` to mention `(pattern)` rather than term (it's not just move, also `elim:` and `case:`).

With hoist I meant that I'd like to see `[]` be part of rewrite's syntax, and `()` be part of `set := ` or `move: ()` as in
`set := term | ( pattern )` and `rewrite [ pattern ]`. IIRC, since the patterns accepted by rewrite are more numerous, we may have a non terminal `rewrite_pattern := pattern | rewrite-specific-patterns-starting-with-in` and use `rewrite_pattern` between `[]`. I just wanted to avoid duplication between the `()` patterns and the `[]` ones, since they are the same (as in one thing to learn).